File: languages/js/src/entity/authenticity/signature_jws.ts
import * as proto from "../../bridge/proto/identity_entities";
import { SignatureAlg } from "./signature_alg";
import { SignatureHeaderJws } from "./signature_header_jws";

/**
 * Represents a JSON Web Signature (JWS). [RFC 7515](https://datatracker.ietf.org/doc/html/rfc7515).
 */
export class SignatureJws {
  signature: string;
  protected: string;
  header: SignatureHeaderJws;
  messageHash: string;

  /**
   * Constructs a SignatureJws object with the specified parameters.
   * @param messageHash 
   * @param signature 
   * @param prot 
   * @param header 
   */
  constructor(
    messageHash: string,
    signature: string,
    prot: string,
    header: SignatureHeaderJws
  ) {
    this.signature = signature;
    this.protected = prot;
    this.header = header;
    this.messageHash = messageHash;
  }

  static fromProto(s: proto.SignatureJWS): SignatureJws {
    return new SignatureJws(
      s.messageHash,
      s.signature,
      s.protected,
      SignatureHeaderJws.fromProto(s.header!)
    );
  }

  toProto(): proto.SignatureJWS {
    return proto.SignatureJWS.fromPartial({
      signature: this.signature,
      protected: this.protected,
      header: this.header.toProto(),
      messageHash: this.messageHash
    });
  }

  getAlg(): SignatureAlg {
    return SignatureAlg.fromString(this.header.alg);
  }
}

File: languages/go/entity/authenticity/signature_jws.go
package authenticity

import (
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// SignatureJws represents a JSON Web Signature (JWS). [RFC 7515].
//
// [RFC 7515]: https://datatracker.ietf.org/doc/html/rfc7515
type SignatureJws struct {
	Signature   string
	Protected   string
	Header      SignatureHeaderJws
	MessageHash string
}

func NewSignatureJwsFromProto(s *proto.SignatureJWS) SignatureJws {
	if s == nil {
		return SignatureJws{}
	}
	return SignatureJws{
		Signature:   s.Signature,
		Protected:   s.Protected,
		Header:      NewSignatureHeaderJwsFromProto(s.Header),
		MessageHash: s.MessageHash,
	}
}

// GetAlg returns the SignatureAlg based on the algorithm specified in the header.
func (s *SignatureJws) GetAlg() SignatureAlg {
	if alg, ok := SignatureAlgFromProto[s.Header.Alg]; ok {
		return alg
	}
	return UNRECOGNIZED_SIGNATURE_ALG
}

func (s SignatureJws) ToProto() *proto.SignatureJWS {
	return &proto.SignatureJWS{
		Signature:   s.Signature,
		Protected:   s.Protected,
		Header:      s.Header.ToProto(),
		MessageHash: s.MessageHash,
	}
}
